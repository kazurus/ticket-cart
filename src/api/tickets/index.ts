import { mapDestinationsToTickets } from './mappers';
import { destinationsListMock, milePriceMock } from './mockData';
import type { Ticket } from '../../store/tickets/types';

export const getTickets: () => Promise<Ticket[]> = async () =>
  mapDestinationsToTickets(destinationsListMock, milePriceMock);

const tickets = {
  getTickets,
};

export default tickets;
