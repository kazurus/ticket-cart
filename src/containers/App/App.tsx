import React from 'react';
import TicketsContainer from '../TicketsContainer';
import OrdersContainer from '../OrdersContainer';

const App = () => (
  <div>
    <h2>Shopping Cart Example</h2>
    <hr />
    <TicketsContainer />
    <hr />
    <OrdersContainer />
  </div>
);

export default App;
