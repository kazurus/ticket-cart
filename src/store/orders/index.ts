export { ordersActions } from './actions';
export { ordersReducer } from './reducers';
export type { OrdersState } from './reducers';
