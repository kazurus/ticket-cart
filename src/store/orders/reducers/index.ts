import { OrdersActionTypes } from '../actionTypes';
import type { OrdersActions } from '../actionTypes';
import type { OrderId, Order } from '../types';

export type OrdersMapById = Record<OrderId, Order>;

export type OrdersState = {
  entities: {
    byId: OrdersMapById;
    allIds: OrderId[];
  };
};

export const initialState: OrdersState = {
  entities: {
    byId: {},
    allIds: [],
  },
};

// Simple id generation
let nextOrderId = 1;

export function ordersReducer(
  state = initialState,
  action: OrdersActions
): OrdersState {
  switch (action.type) {
    case OrdersActionTypes.ADD_ORDER: {
      const { ticketId } = action.payload;
      const order: Order = {
        // eslint-disable-next-line no-plusplus
        id: nextOrderId++,
        ticketId,
        createdAt: Date.now(),
      };
      const byId = { ...state.entities.byId, [order.id]: order };
      const allIds = state.entities.allIds.concat(order.id);

      return {
        ...state,
        entities: {
          byId,
          allIds,
        },
      };
    }

    case OrdersActionTypes.DELETE_ORDER: {
      const { orderId } = action.payload;
      const { [orderId]: value, ...byId } = state.entities.byId;
      const allIds = state.entities.allIds.filter((id) => id !== orderId);

      return {
        ...state,
        entities: {
          byId,
          allIds,
        },
      };
    }

    default:
      return state;
  }
}
