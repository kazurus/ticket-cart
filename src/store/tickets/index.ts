export { ticketsActions } from './actions';
export { watchLoadTickets } from './sagas';
export { ticketsReducer } from './reducers';
export type { TicketsState } from './reducers';
