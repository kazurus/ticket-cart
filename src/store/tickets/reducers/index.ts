import { TicketsActionTypes, TicketsLoadError } from '../actionTypes';
import type { TicketsAction } from '../actionTypes';
import { TicketId, Ticket } from '../types';

export enum TicketsLoadStatus {
  IDLE = 'IDLE',
  PENDING = 'PENDING',
  READY = 'READY',
  ERROR = 'ERROR',
}

export type TicketsMapById = Record<TicketId, Ticket>;

export type TicketsState = {
  ticketsLoadStatus: TicketsLoadStatus;
  error: TicketsLoadError;
  entities: {
    byId: TicketsMapById;
    allIds: TicketId[];
  };
};

export const initialState: TicketsState = {
  ticketsLoadStatus: TicketsLoadStatus.IDLE,
  error: '',
  entities: {
    byId: {},
    allIds: [],
  },
};

function loadTicketsReducer(
  state: TicketsState,
  action: TicketsAction
): TicketsState {
  switch (action.type) {
    case TicketsActionTypes.FETCH_TICKETS: {
      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.PENDING,
        error: '',
      };
    }

    default:
      return state;
  }
}

function loadTicketsPendingReducer(
  state: TicketsState,
  action: TicketsAction
): TicketsState {
  switch (action.type) {
    case TicketsActionTypes.FETCH_TICKETS_SUCCESS: {
      const { tickets } = action.payload;
      const ticketsMapById = tickets.reduce(
        (ticketsMap, ticket) => ({
          ...ticketsMap,
          [ticket.id]: ticket,
        }),
        {}
      );
      const ticketsIds = tickets.map((ticket) => ticket.id);

      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.READY,
        entities: {
          byId: ticketsMapById,
          allIds: ticketsIds,
        },
      };
    }

    case TicketsActionTypes.FETCH_TICKETS_FAILURE: {
      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.ERROR,
        error: action.payload.error,
      };
    }

    default:
      return state;
  }
}

export function ticketsReducer(
  state: TicketsState = initialState,
  action: TicketsAction
): TicketsState {
  switch (state.ticketsLoadStatus) {
    case TicketsLoadStatus.IDLE:
    case TicketsLoadStatus.ERROR:
      return loadTicketsReducer(state, action);
    case TicketsLoadStatus.PENDING:
      return loadTicketsPendingReducer(state, action);

    default:
      return state;
  }
}
