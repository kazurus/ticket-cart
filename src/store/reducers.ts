import { combineReducers } from 'redux';
import { ticketsReducer } from './tickets';
import type { TicketsState } from './tickets';
import { ordersReducer } from './orders';
import type { OrdersState } from './orders';

export type RootState = {
  tickets: TicketsState;
  orders: OrdersState;
};

export const rootReducer = combineReducers({
  tickets: ticketsReducer,
  orders: ordersReducer,
});
