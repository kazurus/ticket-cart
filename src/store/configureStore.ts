import createSagaMiddleware from 'redux-saga';
import { createStore, applyMiddleware, Middleware, compose } from 'redux';
import { createLogger } from 'redux-logger';

import { rootReducer } from './reducers';
import { rootSaga } from './sagas';

export function configureAppStore() {
  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware];

  const excludeEnv = ['production', 'test'];

  if (!excludeEnv.includes(process.env.NODE_ENV)) {
    middlewares.push(createLogger());
  }

  const composeEnhancer =
    (process.env.NODE_ENV !== 'production' &&
      ((window as any)
        .__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ as typeof compose)) ||
    compose;

  const enhancer = composeEnhancer(applyMiddleware(...middlewares));

  const store = createStore(rootReducer, enhancer);

  sagaMiddleware.run(rootSaga);

  return store;
}
