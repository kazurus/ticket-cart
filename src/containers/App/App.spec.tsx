import React from 'react';
import { Store } from 'redux';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureAppStore } from '../../store/configureStore';
import { ticketsActions } from '../../store/tickets';
import { initialState as initialTicketsState } from '../../store/tickets/reducers';
import { initialState as initialOrdersState } from '../../store/orders/reducers';
import App from './App';

const renderOrdersContainer = (store: Store) =>
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );

jest.mock('redux-saga/effects', () => {
  const operators = jest.requireActual('redux-saga/effects');
  const delay = () => new Promise((res) => setTimeout(res, 0));
  return { ...operators, delay };
});

describe('App component', () => {
  let store: ReturnType<typeof configureAppStore>;
  let component: ReturnType<typeof renderOrdersContainer>;

  beforeEach(() => {
    store = configureAppStore();
    component = renderOrdersContainer(store);
    expect(store.getState().tickets).toEqual(initialTicketsState);
    expect(store.getState().orders).toEqual(initialOrdersState);
    store.dispatch(ticketsActions.loadTickets());
  });

  afterEach(() => {
    component.unmount();
  });

  it('should render components', () => {
    const { container } = component;
    expect(container).toMatchSnapshot();
  });

  describe('orders', () => {
    it('should add order', async () => {
      const { queryAllByTestId, findAllByTestId } = component;

      expect(queryAllByTestId('order-item')).toHaveLength(0);

      const [ticketButton] = await findAllByTestId('buy-button');
      fireEvent.click(ticketButton);

      expect(queryAllByTestId('order-item')).toHaveLength(1);
    });

    it('should delete order', async () => {
      const { queryAllByTestId, findAllByTestId } = component;

      const [ticketButton] = await findAllByTestId('buy-button');
      fireEvent.click(ticketButton);

      const [deleteOrderButton] = await findAllByTestId('delete-button');
      fireEvent.click(deleteOrderButton);

      expect(queryAllByTestId('order-item')).toHaveLength(0);
    });
  });
});
