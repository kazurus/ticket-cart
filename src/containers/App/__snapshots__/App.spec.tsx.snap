// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render components 1`] = `
<div>
  <div>
    <h2>
      Shopping Cart Example
    </h2>
    <hr />
    <div>
      <h3>
        Tickets
      </h3>
      <div />
    </div>
    <hr />
    <div>
      <h3>
        Orders
      </h3>
      <div />
    </div>
  </div>
</div>
`;
